### PR TITLE
LRS can reject for its own reasons

### PR DESCRIPTION
**Pull request by garemoko**
Fixes https://github.com/adlnet/xAPI-Spec/issues/575

### CODE REVIEW

File: xAPI.md [hunk 1]
```diff
@@ -2549,6 +2549,15 @@ Agent, and Activity Profile. These four sub-APIs of the Experience API are
 handled via RESTful HTTP methods. The Statement API can be used by itself to 
 track learning records. 
 
+There are certain conditions under which an LRS has to reject statements and other
+conditions which the LRS has to ignore when deciding whether or not to accept a statement. 
+None of this contradicts the idea that the LRS is also able to reject statements for 
+reasons not outlined in this specification. For example, the LRS might assign permissions
+to a particular set of credentials such that those credentials can only issues statements
```
Comment (garemoko) [inline xAPI.md]: typo: "issues" -> "issue"

File: xAPI.md [hunk 2]
```diff
@@ -2549,6 +2549,15 @@ Agent, and Activity Profile. These four sub-APIs of the Experience API are
...
+relating to a particular agent. It could then reject any statements using those credentials
+not relaing to that agent. The permissions that can be assigned by an LRS are out of scope of
```
Comment (garemoko) [inline xAPI.md]: We do talk about permissions within OAuth.

File: xAPI.md [hunk 3]
```diff
@@ -2549,6 +2549,15 @@ Agent, and Activity Profile. These four sub-APIs of the Experience API are
 handled via RESTful HTTP methods. The Statement API can be used by itself to 
 track learning records. 
 
+There are certain conditions under which an LRS has to reject statements and other
+conditions which the LRS has to ignore when deciding whether or not to accept a statement.
```
Comment (garemoko) [inline xAPI.md]: Change "statements" to "requests".

File: xAPI.md [hunk 4]
```diff
@@ -2549,6 +2549,15 @@ Agent, and Activity Profile. These four sub-APIs of the Experience API are
 handled via RESTful HTTP methods. The Statement API can be used by itself to 
 track learning records. 
 
+There are certain conditions under which an LRS has to reject statements and other
+conditions which the LRS has to ignore when deciding whether or not to accept a statement. 
+None of this contradicts the idea that the LRS is also able to reject statements for 
+reasons not outlined in this specification. For example, the LRS might assign permissions
```
Comment (garemoko) [inline xAPI.md]: See error code 403.

File: xAPI.md [hunk 5]
```diff
@@ -2549,6 +2549,15 @@ Agent, and Activity Profile. These four sub-APIs of the Experience API are
 handled via RESTful HTTP methods. The Statement API can be used by itself to 
 track learning records. 
 
+There are certain conditions under which an LRS has to reject statements and other
+conditions which the LRS has to ignore when deciding whether or not to accept a statement. 
+None of this contradicts the idea that the LRS is also able to reject statements for
```
Comment (garemoko) [inline xAPI.md]: LRS should be configurable to pass the conformance tests - think about wording.

**Comment by garemoko** [comment]
Updated following feedback on call. Please review, especially the new requirements and the paragraph starting on [line 2564](https://github.com/adlnet/xAPI-Spec/pull/584/files#diff-68cd6732e400b6e64376effdb84b970eR2564).

**Comment by garemoko** [comment]
I've just seen that we already say ```The LRS MUST handle making, or delegating, decisions on the validity of Statements, and determining what operations might be performed based on the credentials used.``` so this concept is certainly not new.

**Comment by garemoko** [comment]
That has got me wondering if the text I'm adding in 7.0 in fact belongs in 6.4 , security.

**Comment by andyjohnson** [comment]
I wouldn't worry about the location because this is going to be moved to some section more closely tied to LRS behavior when we re-align the document.

**Comment by andyjohnson** [comment]
I'd be curious how others feel about it, but I'd like to change the intro a bit.  Maybe something like:

"This document defines requirements, some of which are imposed on the LRS to reject requests, display query results, and perform other behaviors.  These requirements can be ignored in certain cases not covered in this specification. One of these cases which is out of scope, is that of permissions. "The LRS might assign permissions..."

I'd also like to add something about an LRS choosing to ignore requirements based on malicious intent or overriding system level functions (i.e. Statement is too large).

**Comment by garemoko** [comment]
I don't think it's that requirements can be ignored, it's that the LRS can impose additional restrictions; it can't be less restrictive. 

+1 to ensuring we're clear about statement/document size and malicious intent - I'll have a look at what we already say about statement size.

**Comment by garemoko** [comment]
Ok, I've had a go at that. I've also moved all of this into section 7.1 and added an additional error code to the list. 

The main difference between what I've written and @andyjohnson's suggested wording (much of which I've included) is that I've kept this focused on rejecting requests. The LRS can't ignore requirements even  in the malicious intent scenario; the burden is on us to ensure none of the requirements force the LRS to be insecure.

**Comment by andyjohnson** [comment]
@garemoko , ignore was a poor choice, so thanks for working with it.  What I was trying to get at was that permissions also drive other behaviors, such as the information returned from a request.  Thus a "requirement" of returning All Statements would obscure results from the searcher without sufficient permissions.  There may be other aspects of the spec that are affected by permissions that I'm not currently thinking of as well.

**Comment by garemoko** [comment]
Ok, I see what you're saying now. Makes sense. 

I'll have a think about whether I agree that a blank approach is the right way to go or whether we should explicitly call out hiding certain results based on permissions. The LRS can't partially accept statements or documents based on permissions, so I think returning data is the only instance that can be affected other than fully rejecting requests. 

Permissions around voiding is an area to look at too, and we need to check that we explicitly forbid partially updating documents (I'm thinking in the case of a JSON object, in theory an LRS could give permissions to edit one property and not another - I don't think that's a use case we should support).

**Comment by garemoko** [comment]
There's also authority where in most cases the LRS will overwrite this but not all depending on how much it trusts the credentials, so yeh, I think some blanket-y is best.

**Comment by garemoko** [comment]
I've made those changes and also added a requirement in 7.2 explicitly allowing the LRS to additionally filter statements based on permissions (this is implied by the OAuth "mine" scope, I think). You're look at [line 2557](https://github.com/adlnet/xAPI-Spec/pull/584/files#diff-68cd6732e400b6e64376effdb84b970eR2577).

Note the use of "In cases explicitly allowed by this specification" - I don't think we want to say that the LRS can behave differently based on permissions in cases not covered somewhere in the spec. I think we can be more flexible with applying additional filters when responding to GET requests and with rejecting requests.

**Comment by fugu13** [comment]
It can't be restricted like that; the spec is broadly written to allow things such as additional authentication schemes. If that restriction was in, then LRSs would never be able to restrict permissions for anything but OAuth 3-legged (since that's the only form of restriction allowed by the spec). Of course, LRSs would ignore that, but still, we shouldn't write it in.

**Comment by garemoko** [comment]
'Behave different' is bad wording in my comment. The actual wording I'm referring to is: ```In cases explictly allowed by this specificiation, the credentials used can also affect the LRS behaviour in handling a request, for example the LRS will normally overwrite the Authority property of a Statement, but can sometimes accept a submitted authority if it has a strong trust relationship associated with the credentials used to submit the statement. See [Section 4.1.9 Authority](#authority) for details. ```

note two typos in here I need to fix. 'explictly' 'specificiation'

File: xAPI.md [hunk 6]
```diff
@@ -2618,6 +2645,46 @@ unexpected exception in processing on the server.
...
+* The LRS MUST reject with ```HTTP 403 Forbidden``` status any request rejected by the
+LRS where the credentials associated with the request do not have permission to make that request. 
+
+* The LRS MUST reject with ```HTTP 413 Request Entity Too Large``` status any request rejected by the
+LRS where the size of the Attachment, Statement or Document is larger than the maximum allowed by the LRS.
+
+* The LRS MAY choose any Attachment, Statement and Document size limits and MAY vary this limit on any basis, e.g., per authority.
+
+* The LRS MUST reject with ```429 Too Many Requests``` status any request rejected by the
+LRS where the request is rejected due to too many requests being recieved by a particular client
```
Comment (fugu13) [inline xAPI.md]: very minor typo, "recieved"

File: xAPI.md [hunk 7]
```diff
@@ -2618,6 +2645,46 @@ unexpected exception in processing on the server.
...
+by the LRS do not affect the running of a conformance test suite. 
+
+* The LRS SHOULD* be configurable not to reject any requests from a particular set of credentials on the basis of permissions. 
+This set of credentials SHOULD* be used for conformance testing but MAY be deleted/deactivated on live systems. 
+
+* The LRS MUST be configurable to accept Attachments, Statements or Documents of any reasonable size (see above).
+
+* The LRS MUST be configurable to accept requests at any reasonable rate. 
+
+* In order to test that the LRS returns the correct error codes, the LRS SHOULD* be configurable with three sets of credentials: one which
```
Comment (fugu13) [inline xAPI.md]: I don't really see any value to requiring these; they wouldn't check the actual behavior as they'd be totally special-cased. If they don't reflect the actual case, they aren't useful.

File: xAPI.md [hunk 8]
```diff
@@ -2877,6 +2944,9 @@ which contain both statementId and voidedStatementId parameters
 which contain statementId or voidedStatementId parameters, and also contain any 
 other parameter besides "attachments" or "format".
 
+* The LRS MAY apply additional query filter criteria based on permissions associated
```
Comment (fugu13) [inline xAPI.md]: I like this line, it's very important

File: xAPI.md [hunk 9]
```diff
@@ -2618,6 +2645,46 @@ unexpected exception in processing on the server.
...
+* The LRS SHOULD* be configurable not to reject any requests from a particular set of credentials on the basis of permissions. 
+This set of credentials SHOULD* be used for conformance testing but MAY be deleted/deactivated on live systems. 
+
+* The LRS MUST be configurable to accept Attachments, Statements or Documents of any reasonable size (see above).
+
+* The LRS MUST be configurable to accept requests at any reasonable rate. 
+
+* In order to test that the LRS returns the correct error codes, the LRS SHOULD* be configurable with three sets of credentials: one which
+rejects all requests based on permissions, one which rejects all requests based on size and one which rejects all requests after the first based on rate. 
+These sets of credentials SHOULD* be available within conformance testing but MAY be deleted/deactivated on live systems.
```
Comment (garemoko) [inline xAPI.md]: From the call: delete this requirement.

**Comment by andyjohnson** [comment]
And otherwise we are good save a typo

File: xAPI.md (hunk 6 above)
Comment (garemoko) [inline xAPI.md]: I always forget the "except after c" part of the rhyme.

**Comment by garemoko** [comment]
Fixes discussed on the call made. Ready for +1 and merge.

File: xAPI.md [hunk 10]
```diff
@@ -2605,9 +2630,11 @@ a precondition posted with the request, in the case of State or Agent Profile or
 API calls. See Section [6.3 Concurrency](#concurrency) for more details.
 
 * ```413 Request Entity Too Large``` - Indicates that the LRS has rejected the Statement or 
-document because its size is larger than the maximum allowed by the LRS. The LRS is free to
-choose any limit and MAY vary this limit on any basis, e.g., per authority, but
-MUST be configurable to accept Statements of any size.
+Document because its size (or the size of an Attachment included in the request) is larger than 
+the maximum allowed by the LRS. 
+
+* ```Too Many Requests``` - Indicates that the LRS has rejected the request because it has received
```
Comment (ljwolford) [inline xAPI.md]: Should have code 429 for Too Many Requests bullet, yeah?

Comment (garemoko) [inline xAPI.md]: Yep. Thanks! Fixing..

**Pull request merged**